core/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryStorage.java

```diff
@@ -0,0 +1,154 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.telemetry;
+
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.apache.commons.codec.binary.Base64;
+
+import static java.time.temporal.ChronoUnit.DAYS;
+
+/**
+ * Holds all data related to telemetry that needs to be persisted. Provides methods to serialize and deserialize to/from files.
+ */
+class TelemetryStorage {
+  private LocalDate installDate;
+  private LocalDate lastUseDate;
+  private LocalDateTime lastUploadDateTime;
+  private Long numUseDays;
+  private boolean enabled;
+
+  public TelemetryStorage() {
+    enabled = true;
+    installDate = LocalDate.now();
+    lastUseDate = null;
+    lastUploadDateTime = null;
+    numUseDays = 0L;
+  }
+
+  TelemetryStorage setLastUploadTime(LocalDateTime dateTime) {
+    this.lastUploadDateTime = dateTime;
+    return this;
+  }
+
+  public TelemetryStorage setEnabled(boolean enabled) {
+    this.enabled = enabled;
+    return this;
+  }
+
+  TelemetryStorage setInstallDate(LocalDate installDate) {
+    this.installDate = installDate;
+    return this;
+  }
+
+  TelemetryStorage setLastUseDate(LocalDate lastUseDate) {
+    this.lastUseDate = lastUseDate;
+    return this;
+  }
+
+  TelemetryStorage setNumUseDays(Long numUseDays) {
+    this.numUseDays = numUseDays;
+    return this;
+  }
+
+  public boolean enabled() {
+    return enabled;
+  }
+
+  @CheckForNull
+  public LocalDateTime lastUploadDateTime() {
+    return lastUploadDateTime;
+  }
+
+  public LocalDate installDate() {
+    return installDate;
+  }
+
+  @CheckForNull
+  public LocalDate lastUseDate() {
+    return lastUseDate;
+  }
+
+  public Long numUseDays() {
+    return numUseDays;
+  }
+
+  static TelemetryStorage load(Path filePath) throws IOException {
+    Gson gson = new Gson();
+    byte[] bytes = Files.readAllBytes(filePath);
+    byte[] decoded = Base64.decodeBase64(bytes);
+    String json = new String(decoded, StandardCharsets.UTF_8);
+    return validate(gson.fromJson(json, TelemetryStorage.class));
+  }
+
+  public void save(Path filePath) throws IOException {
+    Files.createDirectories(filePath.getParent());
+    Gson gson = new Gson();
+    String json = gson.toJson(this);
+    byte[] encoded = Base64.encodeBase64(json.getBytes(StandardCharsets.UTF_8));
+    Files.write(filePath, encoded);
+  }
+  
+  public void safeSave(Path filePath) {
+    try {
+      save(filePath);
+    } catch (IOException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)